File: 2015/07/parse.py
from parseInstruction import parseInstruction


def parse(input):
    parsed = {}
    for line in input.split("\n"):
        instruction = parseInstruction(line)
        if instruction["dest"] in parsed:
            raise Exception(
                "Wire '%s' cannot recieve signal from multiple sources"
                % instruction["dest"]
            )
        parsed[instruction["dest"]] = {
            "gate": instruction["gate"],
            "a": instruction["a"],
            "status": "NOT_CALCULATED",
        }
        if instruction["gate"] in ["AND", "OR", "LSHIFT", "RSHIFT"]:
            parsed[instruction["dest"]]["b"] = instruction["b"]
    return parsed

File: 2015/07/parse.js
const parseInstruction = require("./parseInstruction");

function parse(contents) {
  let instructions = {};
  for (let line of contents.split("\n")) {
    let instruction = parseInstruction(line);
    if (instructions[instruction.destination] !== undefined) {
      throw new Error("a wire can only recieve signal from one source");
    }
    instructions[instruction.destination] = {
      gate: instruction.gate,
      a: instruction.a,
      b: instruction?.b,
    };
  }
  return instructions;
}

module.exports = parse;
